Check L3 agent mode before migration

stefannica opened the pull request:
The neutron-ha-tool --l3-agent-migrate and --l3-agent-evacuate
options now also check the mode (dvr/dvr_snat/legacy) configured
for the L3 agents to ensure that routers are migrated only between
L3 agents with the same mode. In addition to that, L3 agents
configured in a dvr-only mode are now ignored by the neutron-ha-tool,
since routers hosted by them cannot be migrated.
This corrects neutron-ha-tool's L3 agent mismatch errors seen in
DVR HA setups, such as migrating a distributed router from/to a dvr
mode L3 agent.

stefannica commented:
![](https://github.trello.services/images/mini-trello-icon.png) [P2: Bug 1046857 - #L3 In HA setup not all ports / namespaces come up after the reboot of a node](https://trello.com/c/84t68MQO/340-p2-bug-1046857-l3-in-ha-setup-not-all-ports-namespaces-come-up-after-the-reboot-of-a-node)

stefannica commented:
Forgot to add some tests for these DVR scenarios

stefannica commented:
I tested this on an HA/DVR setup by having a VM pinging the internet and using three approaches to trigger the neutron-ha-tool:
- manually invoking 'neutron-ha-tool --l3-agent-evacuate'
- putting one controller node in maintenance mode and stopping the l3 agent on it
- powering off one controller node

In all cases, the routers have been migrated to another controller and after a brief pause, the outside VM traffic was successfully resumed.

jsuchome commented:
Could you test it in the Cloud6-7 upgrade scenario?

stefannica commented:
I updated the fix to completely ignore offline dvr-only L3 agents. The previous version only removed the routers from offline dvr-only L3 agents, but had another undesired effect: when powering off the main controller node, sometimes the neutron-ha-tool finds all L3 agents offline except one (probably the one running on the other controller node), which leads to the removal of routers from all dvr-only L3 agents, which practically disrupts all external access for all VMs.
The new approach is to leave offline dvr-only L3 agents untouched. The fact that they still show up in the list of L3 agents hosting a router is harmless (doesn't eat up any resources) and the operation of migrating routers to another dvr-only L3 agent is a nonsense.

stefannica commented:
@jsuchome, the Cloud7 Upgrade HA/DVR job passed. There was just one tempest test that failed during the job execution, but then subsequently passed when I re-executed tempest manually - probably a parallel execution timing fluke.

jsuchome commented:
@stefannica Nice, thanks a lot!

Now, I won't block this review, but I also do not feel confident regarding this code to approve it ... hopefully @aplanas or @matelakat can do the review...

matelakat commented on files/default/test-neutron-ha-tool.py:
```diff
@@ -162,7 +239,7 @@ def test_evacuate_without_agents_returns_no_errors(self):
         )
 
-        self.assertEqual(0, error_count)
+        self.assertEqual(1, error_count)
```
Why did it change?

stefannica replied:
My mistake. I reverted it to the previous behaviour.

matelakat commented on files/default/test-neutron-ha-tool.py:
```diff
@@ -85,32 +90,40 @@ def list_floatingips(self, router_id):
 
 
-def setup_fake_neutron(live_agents=0, dead_agents=0):
+def setup_fake_neutron(live_agents=0, dead_agents=0,
```
I would amend this method instead of adding a new one. The setup code is not really pleasant to read. It would be better if we could say:
```python
setup_fake_neutron(2*[DeadAgent()], 3*[LiveAgent()], 1*[LiveAgent(mode='dvr')])
```
Instead of using update_fake_neutron. Mutating too much these fakes is not a good idea either. Also `update` is a quite generic word to use.

stefannica replied:
Agreed. I refactored the tests to do something similar to your suggestion.

matelakat commented on files/default/test-neutron-ha-tool.py:
```diff
@@ -150,10 +163,74 @@ def test_migrate_from_dead_agent_moves_routers_and_returns_no_errors(self):
+    def test_migrate_from_dead_dvr_snat_agent_moves_routers_and_returns_no_errors(self):
+        fake_neutron = setup_fake_neutron(live_agents=1, dead_agents=1,
+                                          agent_mode='dvr_snat')
+        update_fake_neutron(fake_neutron, live_agents=2, agent_mode='dvr')
```
You can see that this above doesn't read nicely. You have to look at the code for `update_fake_neutron` to understand what is the starting state. See my previous comment.

stefannica replied:
That's right, it looks much nicer with the new mode.

matelakat commented on files/default/neutron-ha-tool.py:
```diff
@@ -943,45 +970,43 @@ def list_alive_agents(agent_list, agent_type):
             if agent['agent_type'] == agent_type and
             agent['alive'] and
-            agent['host'] != exclude_agent_host and
+            agent['admin_state_up'] is True and
```
This `admin_state_up` is a separate fix, isn't it?

stefannica replied:
Yes, it is a separate fix, but with the latest update completely replacing the target_agent_list with the list_alive_l3_agents, this fix is completely factored into the changeset, so it's impossible to detach from it.

matelakat commented on files/default/neutron-ha-tool.py:
```diff
@@ -488,11 +496,13 @@ def l3_agent_migrate(qclient, agent_picker, router_filter, noop=False,
 
     timeout = 0
     if not now:
+        agent_dead_list_old = agent_dead_list
```
why is this change? Is it a refactoring step?

stefannica replied:
Yes, it was a refactoring change. It wasn't necessary and I removed it.

matelakat reviewed (changes requested):
I think this PR contains several Fixes, not just one. We would need to separate those changes to separate commits, otherwise it's hard to understand. For example, there is a refactoring step, which I think is not needed. There is also adding an extra filter on `admin_state_up` which might need a bit of explanation. Also, I think neutron-ha-tool does not want to deal with dvr routers, so why don't we filter out them on the lowest level? see that all the callers of `list_dead_agents` have their `exclude_agent_mode` set to dvr. Also the tests would need some refactoring - I included some suggestions on that as well, but I can live with them as they are, but I expect each change to have a separate change, which we don't have at the moment (see admin_state_up modification)

stefannica commented:
@matelakat, I removed the refactoring bit, filtered dvr l3 agents globally and updated the unittests as you suggested. However, as a result, the 'admin_state_up' fix is no longer obvious and cannot be detached in a separate commit. What do you suggest I do ?

stefannica committed:
Filter out admin disabled l3 agents as evac targets

When evacuating an l3 agent, other administratively
disabled l3 agents should not be considered as potential
targets.

stefannica commented:
@matelakat, please disregard my previous question. I found a way to separate the changes.

matelakat reviewed (changes requested):
@stefannica This change is still a mix of a refactorings and fixes:
 - refactor: use list_alive_l3_agents
 - refactor: list_dead_l3_agents
 - fix: fix comments of migrate_router
 - fix: exclude dvr routers
 
I would prefer if we didn't put the responsibility of filtering out an agent to the list_alive_agents, let's leave a separate function for that. It reads very badly: "Give me all the agents, expect this one"

Also in the tests the setup_fake_agent is not quite descriptive, so it's hard to tell that this line will set up 2 live agents:
```python
fake_neutron = setup_fake_neutron(2*[setup_fake_agent()])
```

matelakat commented:
@stefannica Please add a test for the admin_state_up change as well to the same commit

stefannica committed:
Refactoring: list_dead_agents and list_alive_agents

Refactor list_dead_agents into list_dead_l3_agents
and list_alive_agents into list_alive_l3_agents as
they were previously only used for l3 agents.

stefannica committed:
Fix migrate_router function doc

stefannica committed:
Refactor target_agent_list to accept exclude_agent param

Refactoring the target_agent_list function to accept
an exclude_agent parameter (agent dict) instead of the
exclude_agent_host parameter (host name).

stefannica committed:
Ignore dvr mode L3 agents in neutron-ha-tool

L3 agents configured in a dvr mode are now ignored by
neutron-ha-tool, since routers cannot be migrated away from or
to dvr mode l3 agents.
This corrects neutron-ha-tool's L3 agent mismatch errors seen in
DVR HA setups, such as migrating a distributed router from a dvr_snat
to a dvr mode L3 agent or vice-versa.
Filtering out dvr mode l3 agents also ensures that routers are
migrated between l3 agents configured with the same mode, since
all remaining l3 agents will be of the same mode.

stefannica committed:
Refactor target_agent_list to reuse list_alive_l3_agents

Refactoring the target_agent_list function to reduce
code duplication and reuse the list_alive_l3_agents function
which provides almost the same functionality.

NOTE: The agent mode doesn't need to be explicitly matched
between the migration source and target l3 agents because
they are implicitly of the same mode by filtering out
dvr-only l3 agents.

stefannica committed:
Updated test-neutron-ha-tool with l3 agent mode support

stefannica committed:
Added unit tests for HA DVR scenarios

stefannica committed:
Unit-test - filtering out admin disabled l3 agents as evac targets

stefannica commented:
@matelakat, that's the best I can do wrt separating changes in individual commits. A bit overkill IMO, but here it is nevertheless.

matelakat reviewed (approved):
@stefannica Thank you for your effort, it was much easier to review the change.
